db-engine/src/main/java/com/splicemachine/db/impl/ast/QueryRewriteVisitor.java

```diff
@@ -0,0 +1,242 @@
+        // the distinct sort would not eliminate.
+        if (selectNode.getResultColumns() != null) {
+            CollectNodesVisitor cnv = new CollectNodesVisitor(MethodCallNode.class);
+            selectNode.getResultColumns().accept(cnv);
+            List<MethodCallNode> methodList = cnv.getList();
+            for (MethodCallNode methodCallNode : methodList) {
+                if (!methodCallNode.isDeterministic())
+                    return;
+            }
+        }
+
+        FromList fromList = resultSet.getFromList();
+        for (int i = fromList.size() - 1; i >= 0; i--) {
+            FromTable ft = (FromTable) (fromList.elementAt(i));
+
+            // TODO:  If the inner table of left or right outer join
+            //        has no referenced columns, replace the join
+            //        with the outer table.
+            //        Something like this:
+            /**
+            if (ft instanceof HalfOuterJoinNode) {
+                HalfOuterJoinNode hoj = (HalfOuterJoinNode) ft;
+                ResultSetNode innerTable, outerTable;
+                if (hoj.isRightOuterJoin()) {
+                    innerTable = hoj.getLeftResultSet();
+                    outerTable = hoj.getRightResultSet();
+                }
+                else {
+                    innerTable = hoj.getRightResultSet();
+                    outerTable = hoj.getLeftResultSet();
+                }
+                if (!innerTable.getResultColumns().hasReferencedColumn() &&
+                    (outerTable instanceof FromTable)) {
+                    fromList.setElementAt(outerTable, i);
+                    selectNode.bindResultColumns(parent.getFromList());
+                }
+            }
+            **/
+
+            ResultColumnList rcl = ft.getResultColumns();
+            if (!rcl.hasReferencedColumn()) {
+                // Remove the table from the FROM clause and add:
+                // WHERE EXISTS (select 1 from table) to the where clause, if
+                // there is more than one table in the FROM clause, or,
+                // if there is only one table left in the FROM clause, convert
+                // the query to a TOP 1 query.
+                if (fromList.size() == 1) {
+                    NumericConstantNode
+                      numberOne = new NumericConstantNode(selectNode.getContextManager(),
+                                                            C_NodeTypes.INT_CONSTANT_NODE,
+                                                            Integer.valueOf(1));
+                    selectNode.pushFetchFirst(numberOne);
+                }
+                else {
+                    fromList.removeElementAt(i);
+                    constructExistsCondition(selectNode, ft);
+                }
+            }
+        }
+    }
+
+    private void constructExistsCondition(SelectNode selectNode, FromTable fromTable) throws StandardException {
+        ContextManager cm = fromTable.getContextManager();
```

Comment:
Could the exists subquery constructed here be flattened back into from list later by the subquery flattening logic? I think there should be no problem about correctness, but not sure if it's always alright to assume the plan has only subquery/limit but no joins.

Reply:
@ascend1 It's conceivable that the exists could get flattened back into a left outer join.  I also don't see a problem with that, assuming all of the pre-existing flattening rules are correct.  I'm not sure I follow your point about the assumption that there are no joins.  Any chain of outer joins are self-contained in the fromTable parameter here and the Subquery/Limit gets applied on the entire chain of them, for example:

explain
 select * from t3 where false union
select 1,1,1 from (select * from t1 where true) b, t2 a
right outer join t3 b on a2 = a3 and b2 = b3
right outer join t1 c on false ;

Plan
----
Cursor(n=22,rows=19,updateMode=READ_ONLY (1),engine=OLTP (default))
  ->  ScrollInsensitive(n=22,totalCost=12.404,outputRows=19,outputHeapSize=0 B,partitions=1,parallelTasks=1)
    ->  Distinct(n=21,totalCost=4.044,outputRows=19,outputHeapSize=0 B,partitions=1,parallelTasks=1)
      ->  Union(n=19,totalCost=4.044,outputRows=19,outputHeapSize=0 B,partitions=1,parallelTasks=1)
        ->  Limit(n=18,totalCost=4.044,outputRows=18,outputHeapSize=0 B,partitions=1,parallelTasks=1,fetchFirst=1)
          ->  ProjectRestrict(n=17,totalCost=0.238,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1)
            ->  ProjectRestrict(n=16,totalCost=4.044,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1,preds=[is not null(subq=15)])
              ->  Subquery(n=15,totalCost=502.472,outputRows=1,outputHeapSize=1.562 KB,partitions=1,parallelTasks=1,correlated=false,expression=true,invariant=true)
                ->  Limit(n=15,totalCost=502.44,outputRows=1,outputHeapSize=1.562 KB,partitions=1,parallelTasks=1,fetchFirst=1)
                  ->  ProjectRestrict(n=14,totalCost=502.44,outputRows=1,outputHeapSize=1.562 KB,partitions=1,parallelTasks=1)
                    ->  NestedLoopLeftOuterJoin(n=12,totalCost=502.44,outputRows=1,outputHeapSize=1.562 KB,partitions=1,parallelTasks=1)
                      ->  ProjectRestrict(n=11,totalCost=502.44,outputRows=400,outputHeapSize=1.562 KB,partitions=1,parallelTasks=1)
                        ->  BroadcastLeftOuterJoin(n=10,totalCost=12.3,outputRows=20,outputHeapSize=80 B,partitions=1,parallelTasks=1,preds=[(B2[8:2] = B3[6:2]),(A2[8:1] = A3[6:1])])
                          ->  TableScan[T2(1712)](n=8,totalCost=4.04,scannedRows=20,outputRows=20,outputHeapSize=80 B,partitions=1,parallelTasks=1)
                          ->  TableScan[T3(1728)](n=6,totalCost=4.04,scannedRows=20,outputRows=20,outputHeapSize=40 B,partitions=1,parallelTasks=1)
                      ->  TableScan[T1(1696)](n=4,totalCost=4.04,scannedRows=20,outputRows=20,outputHeapSize=0 B,partitions=1,parallelTasks=1)
              ->  TableScan[T1(1696)](n=3,totalCost=0.224,scannedRows=1,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1)
        ->  ProjectRestrict(n=2,totalCost=0,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1)
          ->  Values(n=0,totalCost=0,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1)
Table statistics are missing or skipped for the following tables:
MWS.T1, C, B, A
Column statistics are missing or skipped for the following columns:
MWS.T3.A3, MWS.T3.B3

--------------------------
Perhaps I'm missing the point of the query.  Please clarify.

Reply:
OK the question was not clear. I just wanted to ask if the exist subquery generated here would ever get flattened. It's clear with your answer. Thanks.